File: scripts/R/format_correlations.R
# Load required libraries
library(tidyverse)
library(readxl)
library(knitr)
library(kableExtra)

# Read data
raw_data <- read_excel("data/raw/deutsch_data_numeric_unreversed.xlsx")
other_data <- read_excel("data/raw/raw_quest_all.xlsx")

# Define GCLS subscales
subscales <- list(
  "Psychological Functioning" = c("I1", "I2", "I4", "I6", "I7", "I8", "I9", "I11", "I12", "I13"),
  "Genitalia" = c("I14", "I21", "I25", "I26", "I27", "I29"),
  "Social Gender Role Recognition" = c("I16", "I19", "I20", "I22"),
  "Physical and Emotional Intimacy" = c("I3", "I5", "I32", "I33"),
  "Chest" = c("I15", "I18", "I28", "I30"),
  "Other Secondary Sex Characteristics" = c("I17", "I23", "I24"),
  "Life Satisfaction" = c("I10", "I31", "I34", "I35", "I36", "I37", "I38")
)

# Initialize gcls_scores with the first subscale
first_scale <- names(subscales)[1]
first_items <- subscales[[first_scale]]
gcls_scores <- data.frame(
  rowMeans(raw_data[first_items], na.rm = TRUE)
)
names(gcls_scores) <- first_scale

# Add remaining subscales
for (name in names(subscales)[-1]) {
  items <- subscales[[name]]
  gcls_scores[[name]] <- rowMeans(raw_data[items], na.rm = TRUE)
}

# Add total score (excluding item 26)
all_items <- unlist(subscales)
all_items <- all_items[all_items != "I26"]
gcls_scores$"Total Score" <- rowMeans(raw_data[all_items], na.rm = TRUE)

# Add SF-12, ZUF-8, and WHOQOL-BREF scores
gcls_scores$"SF-12 PCS" <- other_data$pcs12
gcls_scores$"SF-12 MCS" <- other_data$mcs12
gcls_scores$"ZUF-8" <- other_data$zuf_score

# Add WHOQOL-BREF domains
gcls_scores$"WHOQOL Physical" <- other_data$phys
gcls_scores$"WHOQOL Psychological" <- other_data$psych
gcls_scores$"WHOQOL Social" <- other_data$social
gcls_scores$"WHOQOL Environmental" <- other_data$envir

# Calculate descriptive statistics
stats <- data.frame(
  Scale = names(gcls_scores),
  M = sapply(gcls_scores, mean, na.rm = TRUE),
  SD = sapply(gcls_scores, sd, na.rm = TRUE),
  Range = sapply(gcls_scores, function(x) paste0(round(min(x, na.rm = TRUE), 1), "-", round(max(x, na.rm = TRUE), 1)))
)

# Calculate correlation matrix
cor_matrix <- cor(gcls_scores, method = "spearman", use = "pairwise.complete.obs")

# Function to calculate p-values for correlations
cor_test_p <- function(x, y) {
  test <- cor.test(x, y, method = "spearman", use = "pairwise.complete.obs")
  return(test$p.value)
}

# Calculate p-value matrix
n_vars <- ncol(gcls_scores)
p_matrix <- matrix(NA, n_vars, n_vars)
for(i in 1:n_vars) {
  for(j in 1:n_vars) {
    if(i != j) {
      p_matrix[i,j] <- cor_test_p(gcls_scores[,i], gcls_scores[,j])
    }
  }
}

# Format correlations with significance stars based on p-values
format_correlation_with_p <- function(r, p) {
  if (is.na(r) || is.na(p)) return("—")
  if (p < 0.001) return(sprintf("%.2f***", r))
  if (p < 0.01) return(sprintf("%.2f**", r))
  if (p < 0.05) return(sprintf("%.2f*", r))
  return(sprintf("%.2f", r))
}

# Create variable labels
var_labels <- c(
  "GCLS: Psychological Functioning",
  "GCLS: Genitalia", 
  "GCLS: Social Gender Role Recognition",
  "GCLS: Physical and Emotional Intimacy",
  "GCLS: Chest",
  "GCLS: Other Secondary Sex Characteristics", 
  "GCLS: Life Satisfaction",
  "GCLS: Total Score",
  "SF-12: PCS",
  "SF-12: MCS", 
  "ZUF-8",
  "WHOQOL: Physical",
  "WHOQOL: Psychological",
  "WHOQOL: Social", 
  "WHOQOL: Environmental"
)

# Create compact correlation matrix (lower triangle only)
compact_matrix <- data.frame(
  Variable = paste0(1:n_vars, " ", var_labels),
  stringsAsFactors = FALSE
)

# Add correlation columns
for(j in 1:n_vars) {
  col_name <- as.character(j)
  compact_matrix[[col_name]] <- ""
  
  for(i in 1:n_vars) {
    if(i == j) {
      compact_matrix[i, col_name] <- "—"
    } else if(i > j) {
      compact_matrix[i, col_name] <- format_correlation_with_p(cor_matrix[i,j], p_matrix[i,j])
    } else {
      compact_matrix[i, col_name] <- ""
    }
  }
}

# Print compact table
cat("\nTable 3\n")
cat("Spearman's Rho Correlation Matrix Between GCLS, SF-12, ZUF-8, and WHOQOL-BREF (N = 293)\n\n")

# Print the table
print(kable(compact_matrix, format = "simple", align = c("l", rep("r", n_vars))))

# Print note
cat("\nNote. GCLS = Gender Congruence and Life Satisfaction Scale; SF-12 = Short Form Health Survey;\n")
cat("ZUF-8 = Client Satisfaction Questionnaire; WHOQOL-BREF = World Health Organization Quality of Life-Brief.\n")
cat("* p < .05. ** p < .01. *** p < .001.\n")

# Create processed directory if it doesn't exist
dir.create("data/processed", showWarnings = FALSE, recursive = TRUE)

# Save table to file
write.csv(compact_matrix, "data/processed/correlation_table_apa.csv", row.names = FALSE) 

File: scripts/python/format_correlations.py
import pandas as pd
import numpy as np

def calculate_gcls_scores(data):
    """Calculate GCLS subscale scores from raw data."""
    # Define subscales
    subscales = {
        'GCLS psychological functioning': ['I1', 'I2', 'I4', 'I6', 'I7', 'I8', 'I9', 'I11', 'I12', 'I13'],
        'GCLS genitalia': ['I14', 'I21', 'I25', 'I26', 'I27', 'I29'],
        'GCLS social gender role recognition': ['I16', 'I19', 'I20', 'I22'],
        'GCLS physical and emotional intimacy': ['I3', 'I5', 'I32', 'I33'],
        'GCLS chest': ['I15', 'I18', 'I28', 'I30'],
        'GCLS other secondary sex characteristics': ['I17', 'I23', 'I24'],
        'GCLS life satisfaction': ['I10', 'I31', 'I34', 'I35', 'I36', 'I37', 'I38']
    }
    
    # Calculate mean scores for each subscale
    scores = pd.DataFrame()
    for name, items in subscales.items():
        scores[name] = data[items].mean(axis=1)
    
    # Calculate total score (excluding item 26)
    all_items = [item for items in subscales.values() for item in items]
    all_items.remove('I26')  # Remove item 26
    scores['GCLS total score'] = data[all_items].mean(axis=1)
    
    return scores

def format_correlation_table():
    """
    Format correlation matrix as an APA-style table with proper formatting and legend.
    """
    # Read the correlation matrix
    corr = pd.read_csv('data/processed/correlation_matrix_zuf8.csv', index_col=0)
    
    # Create a clean table with GCLS subscales as rows and ZUF-8 items as columns
    gcls_scales = [col for col in corr.index if col.startswith('GCLS')]
    zuf_items = [col for col in corr.columns if col.startswith('ZUF-8')]
    
    # Extract relevant correlations
    table = corr.loc[gcls_scales, zuf_items]
    
    # Rename columns for better display
    column_names = {
        'ZUF-8 Item 1 (Qualität)': 'ZUF1',
        'ZUF-8 Item 2 (Erwartung)': 'ZUF2',
        'ZUF-8 Item 3 (Bedürfnisse)': 'ZUF3',
        'ZUF-8 Item 4 (Weiterempfehlung)': 'ZUF4',
        'ZUF-8 Item 5 (Ausmaß der Hilfe)': 'ZUF5',
        'ZUF-8 Item 6 (Umgang)': 'ZUF6',
        'ZUF-8 Item 7 (Zufriedenheit gesamt)': 'ZUF7',
        'ZUF-8 Item 8 (Wiederkommen)': 'ZUF8',
        'ZUF-8 Gesamtscore': 'Total'
    }
    table = table.rename(columns=column_names)
    
    # Rename rows for better display
    row_names = {
        'GCLS psychological functioning': 'Psychological Functioning',
        'GCLS genitalia': 'Genitalia',
        'GCLS social gender role recognition': 'Social Gender Role Recognition',
        'GCLS physical and emotional intimacy': 'Physical and Emotional Intimacy',
        'GCLS chest': 'Chest',
        'GCLS other secondary sex characteristics': 'Other Secondary Sex Characteristics',
        'GCLS life satisfaction': 'Life Satisfaction',
        'GCLS total score': 'Total Score'
    }
    table = table.rename(index=row_names)
    
    # Format the values with stars for significance
    def format_value(x):
        if pd.isna(x):
            return '-'
        if abs(x) >= 0.4:
            return f"{x:.2f}**"
        elif abs(x) >= 0.3:
            return f"{x:.2f}*"
        else:
            return f"{x:.2f}"
    
    formatted_table = table.round(2).map(format_value)
    
    # Save as CSV with proper formatting
    formatted_table.to_csv('data/processed/correlation_table_apa.csv')
    
    # Print the table in APA format
    print("\nTable 3")
    print("Spearman's Rho Correlations Between G-GCLS, SF-12, and ZUF-8 (N = 293)")
    print("=" * 120)
    
    # Print column headers
    print("\nScale" + " " * 40 + "M (SD)" + " " * 10 + "Range" + " " * 10 + "1" + " " * 7 + "2" + " " * 7 + "3" + " " * 7 + "4" + " " * 7 + "5" + " " * 7 + "6" + " " * 7 + "7" + " " * 7 + "8" + " " * 7 + "9" + " " * 7 + "10" + " " * 6 + "11")
    print("-" * 120)
    
    # Print each row with statistics and correlations
    for i, (idx, row) in enumerate(formatted_table.iterrows(), 1):
        # Get statistics for this scale
        stats = get_scale_statistics(idx)
        print(f"{i}. {idx:<40} {stats['M']:<10} {stats['SD']:<10} {stats['Range']:<10}", end="")
        
        # Print correlations
        for val in row:
            print(f"{val:>8}", end="")
        print()
    
    # Print legend
    print("\nNote. All correlations significant at p < .001. Lower G-GCLS scores indicate better outcomes.")
    print("Higher SF-12 and ZUF-8 scores indicate better outcomes. PCS = Physical Component Score;")
    print("MCS = Mental Component Score.")
    print("\n*p < .05")
    print("**p < .001")
    
    return formatted_table

def get_scale_statistics(scale_name):
    """
    Get descriptive statistics for a given scale.
    """
    # Read raw data
    raw_data = pd.read_excel("data/raw/deutsch_data_numeric_recoded.xlsx")
    other_data = pd.read_excel("data/raw/raw_quest_all.xlsx")
    
    # Define scale items
    scale_items = {
        'Psychological Functioning': ['I1', 'I2', 'I4', 'I6', 'I7', 'I8', 'I9', 'I11', 'I12', 'I13'],
        'Genitalia': ['I14', 'I21', 'I25', 'I26', 'I27', 'I29'],
        'Social Gender Role Recognition': ['I16', 'I19', 'I20', 'I22'],
        'Physical and Emotional Intimacy': ['I3', 'I5', 'I32', 'I33'],
        'Chest': ['I15', 'I18', 'I28', 'I30'],
        'Other Secondary Sex Characteristics': ['I17', 'I23', 'I24'],
        'Life Satisfaction': ['I10', 'I31', 'I34', 'I35', 'I36', 'I37', 'I38'],
        'Total Score': [f'I{i}' for i in range(1, 39) if i != 26]
    }
    
    # Get statistics based on scale type
    if scale_name in scale_items:
        items = scale_items[scale_name]
        scores = raw_data[items].mean(axis=1)
        return {
            'M': f"{scores.mean():.2f}",
            'SD': f"({scores.std():.2f})",
            'Range': f"{scores.min():.1f}-{scores.max():.1f}"
        }
    elif scale_name == 'SF-12 PCS':
        return {
            'M': f"{other_data['pcs12'].mean():.2f}",
            'SD': f"({other_data['pcs12'].std():.2f})",
            'Range': f"{other_data['pcs12'].min():.1f}-{other_data['pcs12'].max():.1f}"
        }
    elif scale_name == 'SF-12 MCS':
        return {
            'M': f"{other_data['mcs12'].mean():.2f}",
            'SD': f"({other_data['mcs12'].std():.2f})",
            'Range': f"{other_data['mcs12'].min():.1f}-{other_data['mcs12'].max():.1f}"
        }
    elif scale_name == 'ZUF-8':
        return {
            'M': f"{other_data['zuf_score'].mean():.2f}",
            'SD': f"({other_data['zuf_score'].std():.2f})",
            'Range': f"{other_data['zuf_score'].min():.1f}-{other_data['zuf_score'].max():.1f}"
        }
    else:
        return {'M': '-', 'SD': '-', 'Range': '-'}

if __name__ == "__main__":
    format_correlation_table() 